### PR TITLE
crimson/osd/backfill_state: add the object to be pushed in the peer missing set of PeeringState

### PR DESCRIPTION
**Pull request by xxhdx1985126**
Fixes: https://tracker.ceph.com/issues/67874
Signed-off-by: Xuehan Xu <xuxuehan@qianxin.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/crimson/osd/backfill_state.cc [hunk 1]
```diff
@@ -260,7 +261,11 @@ BackfillState::Enqueuing::update_on_peers(const hobject_t& check)
         check == primary_bi.begin && check == peer_bi.begin) {
       if(peer_bi.objects.begin()->second != obj_v &&
           backfill_state().progress_tracker->enqueue_push(primary_bi.begin)) {
-        backfill_listener().enqueue_push(primary_bi.begin, obj_v);
+	auto &v = backfills[primary_bi.begin].first;
```
Comment (rzarzynski) [inline src/crimson/osd/backfill_state.cc:264]: `auto& [v, peers] = the pair`?

(Here and in other places as well).

File: src/crimson/osd/backfill_state.cc [hunk 2]
```diff
@@ -271,10 +276,22 @@ BackfillState::Enqueuing::update_on_peers(const hobject_t& check)
...
+	peers.push_back(bt);
       }
     }
   }
+  for (auto &backfill : backfills) {
+    auto &soid = backfill.first;
+    auto &obj_v = backfill.second.first;
+    auto &peers = backfill.second.second;
+    logger().debug("{}: pushing {},{} to {}", __func__, soid, obj_v, peers);
+    peering_state().prepare_backfill_for_missing(soid, obj_v, peers);
```
Comment (rzarzynski) [inline src/crimson/osd/backfill_state.cc:292]: Why to not make this a part of the `BackfillListener::enqueue_push()`?

File: src/crimson/osd/backfill_state.cc (hunk 1 above)
Comment (xxhdx1985126) [reply src/crimson/osd/backfill_state.cc:264]: Ah, yes, will change it.

File: src/crimson/osd/backfill_state.cc (hunk 2 above)
Comment (xxhdx1985126) [reply src/crimson/osd/backfill_state.cc:292]: Yeah, will change it.

**Comment by xxhdx1985126** [comment]
All comments are addressed, please take a look again, thanks:-)

**Comment by Matan-B** [comment]
jenkins test api

File: src/crimson/osd/backfill_state.cc [hunk 3]
```diff
@@ -251,16 +251,21 @@ BackfillState::Enqueuing::update_on_peers(const hobject_t& check)
...
   for (const auto& bt : peering_state().get_backfill_targets()) {
     const auto& peer_bi = backfill_state().peer_backfill_info.at(bt);
 
     // Find all check peers that have the wrong version
     if (const eversion_t& obj_v = primary_bi.objects.begin()->second;
         check == primary_bi.begin && check == peer_bi.begin) {
-      if(peer_bi.objects.begin()->second != obj_v &&
-          backfill_state().progress_tracker->enqueue_push(primary_bi.begin)) {
-        backfill_listener().enqueue_push(primary_bi.begin, obj_v);
+      if(peer_bi.objects.begin()->second != obj_v) {
```
Comment (athanatos) [inline src/crimson/osd/backfill_state.cc:262]: if (

**Review by athanatos — APPROVED** [review]

**Comment by Matan-B** [comment]
jenkins retest this please

**Comment by Matan-B** [comment]
https://pulpito.ceph.com/matan-2024-10-13_10:25:17-crimson-rados-wip-matanb-crimson-testing-oct-10-distro-crimson-smithi/

Failures are 2 infra and one seastore

**Pull request merged**